neotron-bmc-pico/src/pwm.rs

```diff
@@ -0,0 +1,81 @@
+use stm32f0xx_hal::pac::{RCC, TIM14};
+
+#[derive(Debug, Default)]
+#[repr(align(8))]
+pub struct Config {
+	/// Whether the speaker is active
+	duration: u8,
+	/// The PWM ticks
+	ticks: u16,
+	/// The duty cycle (0 - 255)
+	duty_cycle: u8,
+}
+
+impl Config {
+    pub fn duty_cycle(&self) -> u8 {
+        self.duty_cycle
+    }
+
+    pub fn set_duty_cycle(&mut self, duty_cycle: u8) {
+        self.duty_cycle = duty_cycle;
+    }
+
+    pub fn ticks(&self) -> u16 {
+        self.ticks
+    }
+
+    pub fn set_ticks(&mut self, ticks: u16) {
+        self.ticks = ticks;
+    }
+
+    pub fn duration(&self) -> u8 {
+        self.duration
+    }
+
+    pub fn set_duration(&mut self, duration: u8) {
+        self.duration = duration;
+    }
+
+	pub fn setup_ch(&self, tim14: &TIM14) {
+		let rcc = RCC::ptr();
```

Comment:
It's unsound to conjure up an RCC pointer, because another module could do the same thing in an interrupt routine, and then you have a read-modify-write race.

You have to take an `&mut Rcc` object as an argument as a guard. Then it's safe to conjure up an RCC pointer (which you have to do because the HAL's `Rcc` object has no useful API to speak of - it just proves you have exclusive access).

Reply:
Thanks! I'll do that.